### ISSUE TITLE
Some commits not syncing with Gerrit-hosted git repository

### ISSUE THREAD

**Issue by marekweb** [sourcegraph/sourcegraph#16387]
While syncing git revisions from Gerrit, I'm trying to identify why some revisions are syncing and other revisions don't seem to be syncing.

Gerrit hosts git repositories and can be synced on Sourcegraph by adding it as a "Generic git host".

Instead of using branches, Gerrit puts commits that contain changes to be reviewed under the namespace `refs/changes` followed by a change identifier.

For example on the test repo`k8s-gerrit`, only one branch exists, but multiple commits exist under `refs/changes` revisions because multiple changes have been created for code review in Gerrit.

```
git ls-remote
From https://gerrit.sgdev.org/k8s-gerrit
24d53f6b6c4f484c41697c35d5a790d045faf6a6        HEAD
976be7314788979fb5a9dcb23d06b0151e4626f4        refs/changes/01/1/1
fea70804f1ae21cb462bbcb5eef32ebefec622b3        refs/changes/01/1/meta
584a10336bbd70dc107582afd96352dbf4a625e8        refs/changes/02/2/1
5e1262686bc61f7b0f8f3682f8a7da34eb07e2d6        refs/changes/02/2/2
9ebdb4a4cdbe89752121aa66187c1bdcb4c1054a        refs/changes/02/2/meta
24d53f6b6c4f484c41697c35d5a790d045faf6a6        refs/heads/master
```

### Steps to reproduce

Have the [Gerrit repo added as a generic git host](https://k8s.sgdev.org/site-admin/external-services/RXh0ZXJuYWxTZXJ2aWNlOjE1) with configuration:

```json
{
  "url": "https://gerrit.sgdev.org",
  "repos": ["k8s-gerrit", "src-cli"]
}
```

Open revision [`refs/changes/01/1/1` on Sourcegraph](https://k8s.sgdev.org/gerrit.sgdev.org/k8s-gerrit@refs/changes/01/1/1)
Open revision [`refs/changes/02/2/1` on Sourcegraph](https://k8s.sgdev.org/gerrit.sgdev.org/k8s-gerrit@refs/changes/02/2/1)

### Expected behavior

Both revision links open the repo at the given revision on Sourcegraph

### Actual behavior

The first link opens, and the second link shows a 404.

The second link to revision `refs/changes/02/2/1` should be equivalent to commit [`584a103`](https://k8s.sgdev.org/gerrit.sgdev.org/k8s-gerrit@584a10336bbd70dc107582afd96352dbf4a625e8) but that also shows a 404.

**Comment by eseliger** [comment]
Would this also fix https://github.com/sourcegraph/sourcegraph/issues/10117?

**Comment by marekweb** [comment]
@eseliger The fact that Gerrit changes are listed in the list of branches in Sourcegraph is a separate issue. Currently (with the present issue resolved, which I believe it is) it's possible to access these chanegs if you know the name of the revision (e.g. `refs/changes/02/2/1`) - that's sufficient to implement the browser extension's Gerrit integration

If we decide to list Gerrit changes as branches, we need to decide how to treat them since they're technically not branches (which are under `refs/head`) nor tags (`refs/tags`). I don't know if we already do anything special for Github PR revisions under `refs/pull`.

**Comment by eseliger** [comment]
Understood, thanks!

**Comment by tsenart** [comment]
@keegancsmith: Any hints on why gitserver would not correctly fetch new refs when asked to update the repo?

**Comment by keegancsmith** [comment]
I believe the fix is at https://github.com/sourcegraph/sourcegraph/pull/16767 :)

**Comment by marekweb** [comment]
Customer is reporting that not all Gerrit revisions are syncing. We weren't able to reproduce it on our own instance at gerrit.sgdev.org. @sourcegraph/core-application

Could that be a reoccurence of this problem?

See latest state of this issue https://github.com/sourcegraph/customer/issues/206#issuecomment-810546957

**Comment by keegancsmith** [comment]
TL;DR I think our fix of adding `+refs/changes/*:refs/changes/*` to the fetch refspec is insufficient. We probably need specific `git fetch` for a PR. Inspecting synced refs on a large gerrit instance will help confirm.

In theory `+refs/changes/*:refs/changes/*` should fix this problem. However, gerrit is a custom git server which also stores _all_ data in git. They used to store some stuff in a SQL DB, but at some point they moved everything to be in git. So my theory is gerrit won't advertise all refs under `refs/changes` because it is just too expensive to do. You likely need to mention specific refs.

I tried to confirm this by reading some source code. I came across some references to a heuristic which only advertises the last 30 refs under `refs/changes`. We need to validate this in practice, but I don't have the time to do that right now.

If my theory is correct, we will need to add custom support like we did for Phabricator. For Phabricator we fetch diff via phabricator's REST API then create git commits from them. In this case we need to issue the relevant `git fetch` incantation for a specific ref.

This problem may only occur on specific versions of gerrit. So using the same version as the customer experiencing this issue will be important.

**Comment by christinelovett** [comment]
@keegancsmith customer gerrit version is: 3.2.2

**Comment by christinelovett** [comment]
[customer](https://app.hubspot.com/contacts/2762526/company/557690837)  
[customer](https://app.hubspot.com/contacts/2762526/company/1712829184)

**Comment by ryanslade** [comment]
This currently only appears to be affecting one customer right now so we've asked them to try and replicate the issue manually to confirm whether it's an ACL issue: 

https://github.com/sourcegraph/customer/issues/206#issuecomment-909281535

**Comment by dan-mckean** [comment]
No longer impacting the only known impacted customer.
